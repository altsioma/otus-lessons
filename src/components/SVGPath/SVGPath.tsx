import React from "react";
import { GraphDataType } from "../../types/";

type PathType = {
  color: string;
  strokeWidth: number;
  coords: ISVGPath;
  className?: string;
};

export interface ISVGPath {
  offset: { x: number; y: number };
  multiplier: number;
  data: GraphDataType;
}

const toSVGCoordinates = ({ offset, multiplier, data }: ISVGPath): string => {
  const d = [`M ${offset.x * multiplier} ${offset.y * multiplier}`];
  const collection = data.map((section): string => {
    const xSection = section[0] * multiplier;
    const ySection = section[1] * multiplier;
    return `L ${xSection} ${ySection}`;
  });
  return d.concat(collection).join(" ");
};

export const SVGPath: React.FC<PathType> = ({
  color,
  strokeWidth,
  coords,
  className,
}) => (
  <path
    d={toSVGCoordinates(coords)}
    stroke={color}
    strokeWidth={strokeWidth}
    className={className}
    fill="none"
  ></path>
);
