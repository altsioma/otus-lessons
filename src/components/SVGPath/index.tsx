import React from "react";

type PathType = {
  color: string;
  strokeWidth: number;
  coords: string;
};

export const SVGPath: React.FC<PathType> = ({ color, coords, strokeWidth }) => {
  return (
    <path
      d={coords}
      stroke={color}
      strokeWidth={strokeWidth}
      fill="none"
    ></path>
  );
};
