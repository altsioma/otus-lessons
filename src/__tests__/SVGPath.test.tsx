import React from "react";
import { mount, shallow } from "enzyme";

import { SVGPath } from "../components/SVGPath/SVGPath";

describe("SVGPath work correct", () => {
  it("SVGPath work correct with stable data", () => {
    expect(
      mount(
        <svg>
          <SVGPath coords={{ offset: { x: 10, y: 10 }, multiplier: 10, data: [[1, 3]] }} color="black" strokeWidth={2} />
        </svg>
      ).html()
    ).toMatchSnapshot();
  });

  it('svg path element with parameter coords="M 10 10 L 20 40" has the expected attribute d = M 10 10 L 20 40', () => {
    const wrapper = shallow(
      <SVGPath coords={{ offset: { x: 10, y: 10 }, multiplier: 10, data: [[1, 3]] }} color="black" strokeWidth={2} />

    );
    const path = wrapper.find("path");
    expect(path.props()).toHaveProperty("d", "M 10 10 L 20 40");
  });

  it("svg path stroke attribute has correct value", () => {
    const wrapper = shallow(
      <SVGPath coords={{ offset: { x: 0, y: 0 }, multiplier: 1, data: [[0, 0]] }} color="black" strokeWidth={2} />
    );
    const path = wrapper.find("path");
    expect(path.props()).toHaveProperty("stroke", "black");
  });

});
