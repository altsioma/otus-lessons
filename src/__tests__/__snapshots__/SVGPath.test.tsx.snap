// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SVGPath work correct SVGPath work correct with stable data 1`] = `"<svg><path d=\\"M 10 10 L 20 40\\" stroke=\\"black\\" stroke-width=\\"2\\" fill=\\"none\\"></path></svg>"`;
